File: leet code/Bag of Tokens.cpp


/* 
    Solution by Rahul Surana
    
    ***********************************************************


You have an initial power of power, an initial score of 0, and a bag of tokens where tokens[i] is the value of the ith token (0-indexed).

Your goal is to maximize your total score by potentially playing each token in one of two ways:

If your current power is at least tokens[i], you may play the ith token face up, losing tokens[i] power and gaining 1 score.
If your current score is at least 1, you may play the ith token face down, gaining tokens[i] power and losing 1 score.
Each token may be played at most once and in any order. You do not have to play all the tokens.

Return the largest possible score you can achieve after playing any number of tokens.



    ***********************************************************
*/


#include <bits/stdc++.h>

class Solution {
public:
    
    int ans = 0;
    
    int bagOfTokensScore(vector<int>& tokens, int power) {
        sort(tokens.begin(),tokens.end());
        int i=0,j=tokens.size()-1,s=0;
        while(i<=j)
        {
            if(tokens[i] <= power){
                power-=tokens[i++];
                s++;
                ans = max(ans,s);
            }
            else if(s>0){
                power += tokens[j--];
                s--;
            }
            else break;
        }
        return ans;
    }
};

File: leet code/Bag of Tokens.java


/* 
    Solution by Rahul Surana
    
    ***********************************************************


You start with an initial power of power, an initial score of 0, and a bag of tokens given as an integer array tokens, where each tokens[i] donates the value of tokeni.

Your goal is to maximize the total score by strategically playing these tokens. In one move, you can play an unplayed token in one of the two ways (but not both for the same token):

Face-up: If your current power is at least tokens[i], you may play tokeni, losing tokens[i] power and gaining 1 score.
Face-down: If your current score is at least 1, you may play tokeni, gaining tokens[i] power and losing 1 score.
Return the maximum possible score you can achieve after playing any number of tokens.


    ***********************************************************
*/

class Solution {
    public int bagOfTokensScore(int[] tokens, int power) {
        Arrays.sort(tokens);
        int ans = 0, fa = 0, i = 0, j = tokens.length-1;
        while(j>=i){
            if(power >= tokens[i]) { ans += 1; power-=tokens[i++]; fa = Math.max(fa,ans); }
            else if(ans >= 1) { ans-=1; power += tokens[j--]; }
            else return fa;  
        }
        return fa;
    }
}